### PR TITLE
[spirv] Take bitwidth into account when casting

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 1]
```diff
@@ -287,6 +287,148 @@ uint32_t TypeTranslator::getLocationCount(QualType type) {
...
+    case BuiltinType::Short:
+    case BuiltinType::Min12Int: {
+      if (spirvOptions.enable16BitTypes)
+        return 16;
+      else
+        return 32;
+    }
+    // min16uint (ushort) is treated as 16-bit Uint if '-enable-16bit-types'
+    // option is enabled. It is treated as 32-bit Uint otherwise.
+    case BuiltinType::UShort: {
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Oops. yep

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 2]
```diff
@@ -287,6 +287,148 @@ uint32_t TypeTranslator::getLocationCount(QualType type) {
...
+      if (getIntendedLiteralType(ty) != ty) {
+        return getSpirvBitwidth(getIntendedLiteralType(ty));
+      }
+
+      const auto bitwidth = astContext.getIntWidth(type);
+      // All integer variants with bitwidth larger than 32 are represented
+      // as 64-bit int in SPIR-V.
+      // All integer variants with bitwidth of 32 or less are represented as
+      // 32-bit int in SPIR-V.
+      if (type->isSignedIntegerType())
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: oh yes.

File: tools/clang/lib/SPIRV/TypeTranslator.h [hunk 3]
```diff
@@ -125,6 +125,20 @@ class TypeTranslator {
...
+  /// 64 is passed in, the result-id of a SPIR-V vector of size 4 of signed
+  /// 64-bit integers is returned.
+  /// Acceptable bitwidths are 16, 32, and 64.
+  uint32_t getTypeWithCustomBitwidth(QualType type, uint32_t bitwidth);
+
+  /// \brief Returns the realized bitwidth of the given type when represented in
+  /// SPIR-V. Panics if the given type is not a scalar or vector of float or
+  /// integer. In case of vectors, it returns the realized SPIR-V bitwidth of
+  /// the vector elements.
+  uint32_t getSpirvBitwidth(QualType type);
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.h]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 4]
```diff
@@ -363,24 +363,17 @@ uint32_t TypeTranslator::getSpirvBitwidth(QualType type) {
     case BuiltinType::LongLong:
     case BuiltinType::ULongLong:
       return 64;
-    // min16int (short), and min12int are treated as 16-bit Int if
+    // min16int (short), ushort, and min12int are treated as 16-bit integers if
     // '-enable-16bit-types' option is enabled. They are treated as 32-bit
-    // Int otherwise.
+    // integers otherwise.
     case BuiltinType::Short:
+    case BuiltinType::UShort:
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: Oh, sorry. Also `Half` and `Min10Float`?

Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Oh sorry i don't know why i missed that!

**Comment by ehsannas** [comment]
Bots are failing due to reaching the 1-hour limit of Appveyor's free tier. All tests pass locally.

**Pull request merged**